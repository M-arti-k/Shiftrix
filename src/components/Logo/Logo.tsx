import type { FC, JSX } from "react";
import { useEffect, useRef, useState } from "react";
import { Link } from "react-router-dom";
import styles from "./Logo.module.scss";
import clsx from "clsx";

const Logo: FC = (): JSX.Element => {
  const hoursRef = useRef<HTMLDivElement>(null);
  const minutesRef = useRef<HTMLDivElement>(null);
  const [time, setTime] = useState(new Date());

  useEffect(() => {
    const updateTime = (): void => setTime(new Date());

    const intervalId: NodeJS.Timeout = setInterval(updateTime, 60000);
    updateTime();

    return (): void => clearInterval(intervalId);
  }, []);

  useEffect(() => {
    const hours: number = time.getHours() % 12;
    const minutes: number = time.getMinutes();

    const hoursDeg: number = 30 * hours + 0.5 * minutes;
    const minutesDeg: number = 6 * minutes;

    if (hoursRef.current) {
      hoursRef.current.style.transform = `translate(-50%, -50%) rotate(${hoursDeg}deg)`;
    }

    if (minutesRef.current) {
      minutesRef.current.style.transform = `translate(-50%, -50%) rotate(${minutesDeg}deg)`;
    }
  }, [time]);

  return (
    <Link data-testid="logo" to="/" className={styles.logo}>
      <div className={styles["logo-icon"]}>
        <div className={styles["logo-icon__hexagon"]}>
          <div
            data-testid="clock-hours"
            ref={hoursRef}
            className={clsx(
              styles["logo-icon__hours"],
              styles["logo-icon__clock"]
            )}
          ></div>
          <div
            data-testid="clock-minutes"
            ref={minutesRef}
            className={clsx(
              styles["logo-icon__minutes"],
              styles["logo-icon__clock"]
            )}
          ></div>
        </div>
      </div>
      <p className={styles["logo__name"]}>
        Shiftrix <span>v2.0</span>
      </p>
    </Link>
  );
};

export default Logo;
