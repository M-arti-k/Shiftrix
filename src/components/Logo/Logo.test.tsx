import { act, render, screen } from "@testing-library/react";
import Logo from "./Logo";
import { MemoryRouter } from "react-router-dom";

describe("Logo test", (): void => {
  let logo: HTMLElement;
  beforeEach((): void => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date("2025-01-01T03:15:00"));
    render(
      <MemoryRouter>
        <Logo />
      </MemoryRouter>
    );
    logo = screen.getByTestId("logo");
  });
  test("Logo renders", (): void => {
    expect(logo).toBeInTheDocument();
  });
  test("Link leads to the main page", (): void => {
    expect(logo).toHaveAttribute("href", "/");
  });
  test("Display logo text", (): void => {
    expect(screen.getByText(/Shiftrix/i)).toBeInTheDocument();
    expect(screen.getByText(/v2.0/i)).toBeInTheDocument();
  });
  test("Clock displays system`s time", (): void => {
    const clockHours: HTMLElement = screen.getByTestId("clock-hours");
    const clockMinutes: HTMLElement = screen.getByTestId("clock-minutes");

    expect(clockHours).toHaveStyle(
      "transform: translate(-50%, -50%) rotate(97.5deg)"
    );
    expect(clockMinutes).toHaveStyle(
      "transform: translate(-50%, -50%) rotate(90deg)"
    );
  });
  test("Clock updates after one minute", (): void => {
    const clockHours: HTMLElement = screen.getByTestId("clock-hours");
    const clockMinutes: HTMLElement = screen.getByTestId("clock-minutes");

    act((): void => {
      jest.advanceTimersByTime(60000);
    });

    expect(clockHours).toHaveStyle(
      "transform: translate(-50%, -50%) rotate(98deg)"
    );
    expect(clockMinutes).toHaveStyle(
      "transform: translate(-50%, -50%) rotate(96deg)"
    );
  });
  test("Logo`s snapshot", (): void => {
    expect(logo).toMatchSnapshot();
  });
});
