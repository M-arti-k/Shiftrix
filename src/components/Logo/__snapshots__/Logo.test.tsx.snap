// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logo test Logo\`s snapshot 1`] = `
<a
  class="logo"
  data-discover="true"
  data-testid="logo"
  href="/"
>
  <div
    class="logo-icon"
  >
    <div
      class="logo-icon__hexagon"
    >
      <div
        class="logo-icon__hours logo-icon__clock"
        data-testid="clock-hours"
        style="transform: translate(-50%, -50%) rotate(97.5deg);"
      />
      <div
        class="logo-icon__minutes logo-icon__clock"
        data-testid="clock-minutes"
        style="transform: translate(-50%, -50%) rotate(90deg);"
      />
    </div>
  </div>
  <p
    class="logo__name"
  >
    Shiftrix 
    <span>
      v2.0
    </span>
  </p>
</a>
`;
